[ios][android][background-task] add new package: expo-background-task (#33438)
expo/expo: docs/pages/versions/unversioned/sdk/background-task.mdx

---
sidebar_title: BackgroundTask
title: Background Task
description: A library that provides an API for running background tasks.
sourceCodeUrl: 'https://github.com/expo/expo/tree/main/packages/expo-background-task'
packageName: 'expo-background-task'
platforms: ['android', 'ios']
isNew: true
---

import APISection from '~/components/plugins/APISection';
import { APIInstallSection } from '~/components/plugins/InstallSection';
import { Collapsible } from '~/ui/components/Collapsible';
import { PlatformTags } from '~/ui/components/Tag';

> `expo-background-task` provides an API to run deferrable background tasks in a way that optimises battery and power consumption for the end user. This module uses the [`BGTaskScheduler`](https://developer.apple.com/documentation/backgroundtasks/bgtaskscheduler) API on iOS and the [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager) API on Android to schedule tasks. It also uses the [expo-task-manager](task-manager.mdx) Native API for running JavaScript tasks.
> 

Reviewer: ```suggestion
`expo-background-task` provides an API to run deferrable background tasks in a way that optimizes battery and power consumption for the end user. This module uses the [`BGTaskScheduler`](https://developer.apple.com/documentation/backgroundtasks/bgtaskscheduler) API on iOS and the [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager) API on Android to schedule tasks. It also uses the [expo-task-manager](task-manager.mdx) Native API for running JavaScript tasks.

```

## Background tasks

> A background task is a deferrable unit of work that is performed in the background, outside the lifecycle of your app. This is useful for tasks that need to be executed when the app is not active, such as syncing data with a server, fetching new content, or even checking if there are any [Expo Updates](updates.mdx).

Reviewer: ```suggestion
A background task is a deferrable unit of work that is performed in the background, outside the lifecycle of your app. This is useful for tasks that need to be executed when the app is inactive, such as syncing data with a server, fetching new content, or even checking if there are any [Expo Updates](updates.mdx).
```

> A background task is a deferrable unit of work that is performed in the background, outside the lifecycle of your app. This is useful for tasks that need to be executed when the app is inactive, such as syncing data with a server, fetching new content, or even checking if there are any [Expo Updates](updates.mdx).

Reviewer 2: ```suggestion
A background task is a deferrable unit of work that is performed in the background, outside your app's lifecycle. This is useful for tasks that need to be executed when the app is inactive, such as syncing data with a server, fetching new content, or even checking if there are any [`expo-updates`](updates.mdx).
```

- Let's use the Expo package name here instead of the API. Generally, users are more familiar with the package names.
- We can be concise at the end of the first sentence.

### When are background tasks run?

The Expo Background Task API leverages each platform to execute tasks at the most optimal time for both the user and the device.

This means that the task may not run immediately after it is scheduled, but it will run at some point in the future if the system decides so.

A background task will only run if the battery has enough charge (or the device is plugged into power) and the network is available. Without these conditions, the task won't execute. The exact behavior will vary depending on the operating system.

### When will they stop?

> The native platform APIs will stop running background tasks if the user explicitly kills the app (they will be resumed when the user restarts the app) or if the app is uninstalled. If the device is rebooted or the app crashes, the background tasks will be resumed.

Reviewer 3: just wanted to double check - is this accurate? there isn't any way to run a task when the app has been killed?

Author: As discussed in our sync discussion I'll add a bit more info about how this works on Android.

> > **Info** The non-deterministic behavior of background tasks is intentional, aimed at optimizing battery life and enhancing user experience. If a user force stops an app, the system interprets this as a signal not to run background tasks for that app until the user restarts it. On iOS an app is force stopped if the user swipes up on the app in the app switcher. On Android this behaviour depends on the device vendor. Some vendors have implementations where the app is force stopped when the user removes the app from the recent apps list.

Reviewer 2: This is a bit of a large text to call out. Can it be the second paragraph of the same section?

Author: Yes, makes sense. I've refactored the whole section and simplified these sections a bit after some discussions with Brent as well. 

Also refactored the part of the doc about testing (will create a devtools plugin to ease testing).

Reviewer 2: Thank you!

> > **Info** The non-deterministic behavior of background tasks is intentional, aimed at optimizing battery life and enhancing user experience. If a user force stops an app, the system interprets this as a signal not to run background tasks for that app until the user restarts it. On iOS an app is force stopped if the user swipes up on the app in the app switcher. On Android this behaviour depends on the device vendor. Some vendors have implementations where the app is force stopped when the user removes the app from the recent apps list.

Reviewer 2: ```suggestion
> **Info** The non-deterministic behavior of background tasks is intentional, aimed at optimizing battery life and enhancing user experience. If a user force stops an app, the system interprets this as a signal not to run background tasks for that app until the user restarts it.On Android, this behavior depends on the device vendor. On iOS, an app is force stopped if the user swipes up on the app in the app switcher.  Some vendors have implementations where the app is force stopped when the user removes the app from the recent apps list.
```

- Let's apply the platform order preference for Android and iOS.
- We prefer to use American English spellings in our docs [reference: Writing style guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#voice-and-tone).

## Platform differences

### Android&ensp;<PlatformTags platforms={['android']} />

On Android, the [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager) API allows specifying a minimum interval for a task to run (at least 15 minutes). The task will execute sometime after the interval has passed, provided the specified conditions are met.

### iOS&ensp;<PlatformTags platforms={['ios']} />

On iOS, the [`BGTaskScheduler`](https://developer.apple.com/documentation/backgroundtasks/bgtaskscheduler) API decides the best time to launch your background task. The system will consider the battery level, the network availability, and the user's usage patterns to determine when to run the task.

## Known limitations

### iOS&ensp;<PlatformTags platforms={['ios']} />

> The [`Background Tasks`](https://developer.apple.com/documentation/backgroundtasks) API on iOS is not available on simulators, it is only available when running on a physical device.

Reviewer 2: ```suggestion
The [`Background Tasks`](https://developer.apple.com/documentation/backgroundtasks) API is unavailable on iOS simulators. It is only available when running on a physical device.
```

This can be concise.

## Installation

<APIInstallSection />

## Configuration&ensp;<PlatformTags platforms={['ios']} />

To be able to run background tasks on iOS, you need to add the `processing` value to the `UIBackgroundModes` array in your app's **Info.plist** file. This is required for background fetch to work properly.

> **If you're using [CNG](/workflow/continuous-native-generation/)**, the required `UIBackgroundModes` configuration will be applied automatically by prebuild.

Reviewer 2: Can we make this part the first paragraph in this section and then the current first paragraph (`you need to add the `processing`) can follow with the code snippet so we are not repeating the information? Right now, we are repeating the part both in the first paragraph and in the collapsible about adding `processing` value manually.

<Collapsible summary="Configure UIBackgroundModes manually on iOS">

If you're not using Continuous Native Generation ([CNG](/workflow/continuous-native-generation/)), then you'll need to add the following to your **Info.plist** file:

```xml ios/project-name/Supporting/Info.plist
<key>UIBackgroundModes</key>
  <array>
    <string>processing</string>
  </array>
</key>
```

</Collapsible>

## Usage

> Below is an example that demonstrates how to use `expo-background-task`.

Reviewer 2: ```suggestion
Below is an example that demonstrates how to use `expo-background-task`:
```

### Android&ensp;<PlatformTags platforms={['android']} />

On Android, the easiest way to test background tasks is to register a task. When the app is in `debug` mode, a background task should be started as soon as possible, while in release mode, it will wait until the minimum interval has passed before running the task.

> In release mode, the app must be in the background for the background task to run, while in debug mode, the task will run even if the app is in the foreground.

### iOS&ensp;<PlatformTags platforms={['ios']} />

On iOS, testing background tasks is done by [simulating](https://developer.apple.com/documentation/backgroundtasks/starting-and-terminating-tasks-during-development#Launch-a-Task) the launch of the task. This can be done by calling the `triggerTaskWorkerForTestingAsync` method in JavaScript.

Expo Background Task has a convenient helper on iOS that triggers executing the background task worker. This will only work in `debug` mode:

```ts
import * as BackgroundTask from 'expo-background-task';

export const triggerBackgroundTask = async () => {
  await BackgroundTask.triggerTaskWorkerForTestingAsync();
};
```

## Multiple background tasks

> Since the Background Tasks API on iOS and the WorkManager API on Android limit the number of tasks that can be scheduled for a single app, Expo Background Task uses a single worker on both platforms. While you can define multiple JavaScript background tasks, they will all run through this single worker.

Reviewer 2: ```suggestion
Since the WorkManager API on Android and the Background Tasks API on iOS limit the number of tasks that can be scheduled for a single app, Expo Background Task uses a single worker on both platforms. While you can define multiple JavaScript background tasks, they will all run through this single worker.
```
